File: pipeline/src/lsvoxel/chunkpack/voxel_proxy.py
"""voxel_proxy.bin — whole-dataset, one record per OCCUPIED voxel: the count and mean
thumbnail color every chunk's meta.bin already carries per VoxelRecord, lifted into
one small always-resident file.

proxy.bin (one record per chunk slot) gives the volume's silhouette; this gives its
shape at voxel resolution, so the client can draw every voxel of a chunk that is NOT
loaded as a flat mean-colored block (Minecraft's "distant chunks are approximated")
and stream textured chunks only in a tight ring around the camera. The per-voxel
color/count inside meta.bin is exactly what a non-resident chunk hasn't fetched,
hence the copy. Size: 12 B per occupied voxel (bl-160: 14,688 voxels -> ~172 KiB).

Layout (little-endian):
  Header (16B, same size as proxy.bin's):
    magic="LSVV" @0  version=1:u16 @4  reserved:u16 @6  n_voxels:u32 @8
    num_voxels:u16 @12 (world grid, e.g. 160)  voxels_per_chunk:u16 @14 (16)
  VoxelProxyRecord x n_voxels (12B each):
    chunk_id:u32  local_voxel_id:u16  count:u16  color_rgb:u8[3]  flags:u8
  count / color_rgb / flags are byte-for-byte the chunk's VoxelRecord fields. Pack
  construction rejects a count above 65535 rather than silently saturating it.

Records are sorted by (chunk_id, local_voxel_id) ascending — the same order as each
chunk's occupied list (count > 0, local_voxel_id ascending, which is what the
frontend's ChunkLoader builds its instances from) — so a chunk's records are one
contiguous run and the i-th record of that run is instance i of that chunk's mesh.
"""
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

import numpy as np

MAGIC = b"LSVV"
VERSION = 1

VOXEL_PROXY_HEADER_DTYPE = np.dtype(
    [
        ("magic", "S4"),
        ("version", "<u2"),
        ("reserved", "<u2"),
        ("n_voxels", "<u4"),
        ("num_voxels", "<u2"),
        ("voxels_per_chunk", "<u2"),
    ]
)
assert VOXEL_PROXY_HEADER_DTYPE.itemsize == 16, VOXEL_PROXY_HEADER_DTYPE.itemsize

VOXEL_PROXY_RECORD_DTYPE = np.dtype(
    [
        ("chunk_id", "<u4"),
        ("local_voxel_id", "<u2"),
        ("count", "<u2"),
        ("color_rgb", "u1", (3,)),
        ("flags", "u1"),
    ]
)
assert VOXEL_PROXY_RECORD_DTYPE.itemsize == 12, VOXEL_PROXY_RECORD_DTYPE.itemsize
WIDE_PROXY_DTYPE = np.dtype([
    ('chunk_id', '<u4'), ('local_voxel_id', '<u2'), ('count', '<u4'),
    ('color_rgb', 'u1', (3,)), ('flags', 'u1'),
])


@dataclass
class VoxelProxy:
    num_voxels: int
    voxels_per_chunk: int
    records: np.ndarray  # structured array, dtype=VOXEL_PROXY_RECORD_DTYPE

    @property
    def n_voxels(self) -> int:
        return len(self.records)


def records_for_chunk(chunk_id: int, voxel_records: np.ndarray) -> np.ndarray:
    """One chunk's occupied VoxelRecords (metablob.VOXEL_RECORD_DTYPE, index ==
    local_voxel_id) as proxy records, local_voxel_id ascending — exactly the chunk's
    occupied list. Both a fresh build and a derive-from-existing-pack go through
    this, so the two paths cannot disagree about what a record holds."""
    occupied = np.flatnonzero(voxel_records["count"] > 0)
    wide = voxel_records.dtype['count'].itemsize == 4
    recs = np.zeros(len(occupied), dtype=WIDE_PROXY_DTYPE if wide else VOXEL_PROXY_RECORD_DTYPE)
    recs["chunk_id"] = chunk_id
    recs["local_voxel_id"] = occupied
    recs["count"] = voxel_records["count"][occupied]
    recs["color_rgb"] = voxel_records["color_rgb"][occupied]
    recs["flags"] = voxel_records["flags"][occupied]
    return recs


def build_voxel_proxy_records(per_chunk: list[np.ndarray]) -> np.ndarray:
    """Concatenate per-chunk runs (given in any chunk order) into the canonical
    (chunk_id, local_voxel_id)-sorted table."""
    if not per_chunk:
        return np.zeros(0, dtype=VOXEL_PROXY_RECORD_DTYPE)
    recs = np.concatenate(per_chunk)
    if recs.dtype not in (VOXEL_PROXY_RECORD_DTYPE, WIDE_PROXY_DTYPE):
        raise ValueError("per-chunk runs must use VOXEL_PROXY_RECORD_DTYPE")
    order = np.lexsort((recs["local_voxel_id"], recs["chunk_id"]))
    return recs[order]


def check_canonical_order(records: np.ndarray) -> None:
    """Raise unless records are strictly ascending by (chunk_id, local_voxel_id) —
    sorted AND free of duplicate voxels. The composite key fits int64 since
    local_voxel_id < 2**16."""
    key = records["chunk_id"].astype(np.int64) * (1 << 16) + records["local_voxel_id"]
    if len(key) > 1 and not (np.diff(key) > 0).all():
        raise ValueError(
            "records must be sorted by (chunk_id, local_voxel_id) ascending with no duplicates"
        )


def write_voxel_proxy(
    path: Path, num_voxels: int, voxels_per_chunk: int, records: np.ndarray
) -> None:
    if records.dtype not in (VOXEL_PROXY_RECORD_DTYPE, WIDE_PROXY_DTYPE):
        raise ValueError("records must use VOXEL_PROXY_RECORD_DTYPE")
    if voxels_per_chunk <= 0 or num_voxels % voxels_per_chunk != 0:
        raise ValueError(
            f"num_voxels ({num_voxels}) must be divisible by voxels_per_chunk ({voxels_per_chunk})"
        )
    check_canonical_order(records)
    if len(records):
        n_chunk_slots = (num_voxels // voxels_per_chunk) ** 3
        if int(records["chunk_id"].max()) >= n_chunk_slots:
            raise ValueError(f"chunk_id out of range for a {num_voxels}^3 world")
        if int(records["local_voxel_id"].max()) >= voxels_per_chunk**3:
            raise ValueError(f"local_voxel_id out of range for {voxels_per_chunk}^3 voxels per chunk")

    header = np.zeros(1, dtype=VOXEL_PROXY_HEADER_DTYPE)
    header["magic"] = MAGIC
    header["version"] = 2 if records.dtype == WIDE_PROXY_DTYPE else VERSION
    header["n_voxels"] = len(records)
    header["num_voxels"] = num_voxels
    header["voxels_per_chunk"] = voxels_per_chunk

    path.parent.mkdir(parents=True, exist_ok=True)
    with open(path, "wb") as f:
        f.write(header.tobytes())
        f.write(records.tobytes())


def read_voxel_proxy_header(path: Path) -> tuple[int, int, int]:
    """(num_voxels, voxels_per_chunk, n_voxels) from the 16-byte header alone —
    enough to size buffers or fill a manifest entry without reading the records."""
    header = np.fromfile(path, dtype=VOXEL_PROXY_HEADER_DTYPE, count=1)
    if len(header) != 1:
        raise ValueError(f"{path}: truncated header")
    h = header[0]
    if bytes(h["magic"]) != MAGIC:
        raise ValueError(f"{path}: bad magic {bytes(h['magic'])!r}")
    if int(h["version"]) not in (VERSION, 2):
        raise ValueError(f"{path}: unsupported version {h['version']}")
    return int(h["num_voxels"]), int(h["voxels_per_chunk"]), int(h["n_voxels"])


def read_voxel_proxy(path: Path) -> VoxelProxy:
    num_voxels, voxels_per_chunk, n_voxels = read_voxel_proxy_header(path)
    raw = np.fromfile(path, dtype=np.uint8)
    dtype = WIDE_PROXY_DTYPE if int(raw[:16].view(VOXEL_PROXY_HEADER_DTYPE)[0]['version']) == 2 else VOXEL_PROXY_RECORD_DTYPE
    expected_size = VOXEL_PROXY_HEADER_DTYPE.itemsize + n_voxels * dtype.itemsize
    if raw.shape[0] != expected_size:
        raise ValueError(
            f"{path}: size mismatch, file is {raw.shape[0]}B, header implies {expected_size}B"
        )
    records = raw[VOXEL_PROXY_HEADER_DTYPE.itemsize :].view(dtype).copy()
    check_canonical_order(records)
    return VoxelProxy(num_voxels=num_voxels, voxels_per_chunk=voxels_per_chunk, records=records)

File: frontend/src/streaming/VoxelProxy.ts
import { fetchArrayBuffer } from "../net/fetchTyped.ts";
import type { Manifest } from "./Manifest.ts";
import type { VoxelProxyData } from "../types.ts";

const VOXEL_PROXY_MAGIC = "LSVV";
const VOXEL_PROXY_HEADER_BYTES = 16;

/**
 * Decodes `voxel_proxy.bin` — the whole dataset's occupied voxels, each with
 * the count and mean thumbnail color its chunk's `meta.bin` carries, lifted
 * into one small always-resident file (bl-160: 14,688 records, 172 KiB) so the
 * client can draw every voxel of a chunk that is NOT streamed in as a flat
 * mean-colored block. See `pipeline/src/lsvoxel/chunkpack/voxel_proxy.py`.
 *
 * Layout (little-endian — `DataView` defaults to big-endian, so every
 * multi-byte read passes `true`):
 *
 *   header 16B: magic "LSVV" | version u16 | reserved u16 | n_voxels u32
 *               | num_voxels u16 | voxels_per_chunk u16
 *   VoxelProxyRecord[n_voxels], 12B each:
 *               chunk_id u32 | local_voxel_id u16 | count u16
 *               | color_rgb u8[3] | flags u8
 *
 * Records must be strictly ascending by `(chunk_id, local_voxel_id)`; the
 * per-chunk run table (`runStart`/`runEnd`) is built while checking that, and
 * everything downstream — hiding a chunk's run when it becomes resident,
 * resolving a raycast hit back to a voxel — leans on the order being exact.
 */
export function parseVoxelProxy(buffer: ArrayBuffer): VoxelProxyData {
  const view = new DataView(buffer);
  const magic = String.fromCharCode(
    view.getUint8(0),
    view.getUint8(1),
    view.getUint8(2),
    view.getUint8(3),
  );
  if (magic !== VOXEL_PROXY_MAGIC) {
    throw new Error(`voxel_proxy.bin: bad magic ${JSON.stringify(magic)}`);
  }
  const version = view.getUint16(4, true);
  if (version !== 1 && version !== 2) throw new Error(`voxel_proxy.bin: unsupported version ${version}`);
  const recordBytes = version === 1 ? 12 : 14;
  const nVoxels = view.getUint32(8, true);
  const numVoxels = view.getUint16(12, true);
  const voxelsPerChunk = view.getUint16(14, true);

  const expectedBytes = VOXEL_PROXY_HEADER_BYTES + nVoxels * recordBytes;
  if (buffer.byteLength !== expectedBytes) {
    throw new Error(
      `voxel_proxy.bin: size mismatch — got ${buffer.byteLength}B, header implies ${expectedBytes}B`,
    );
  }
  if (voxelsPerChunk === 0 || numVoxels % voxelsPerChunk !== 0) {
    throw new Error(
      `voxel_proxy.bin: num_voxels ${numVoxels} is not a multiple of voxels_per_chunk ${voxelsPerChunk}`,
    );
  }
  const chunksPerAxis = numVoxels / voxelsPerChunk;
  const chunkSlots = chunksPerAxis ** 3;
  const voxelsPerChunkTotal = voxelsPerChunk ** 3;

  const chunkId = new Uint32Array(nVoxels);
  const localVoxelId = new Uint16Array(nVoxels);
  const count = new Uint32Array(nVoxels);
  const colorRgb = new Uint8Array(nVoxels * 3);
  const flags = new Uint8Array(nVoxels);
  const runStart = new Int32Array(chunkSlots).fill(-1);
  const runEnd = new Int32Array(chunkSlots);

  let previousKey = -1;
  for (let i = 0; i < nVoxels; i++) {
    const base = VOXEL_PROXY_HEADER_BYTES + i * recordBytes;
    const c = view.getUint32(base, true);
    const v = view.getUint16(base + 4, true);
    if (c >= chunkSlots || v >= voxelsPerChunkTotal) {
      throw new Error(`voxel_proxy.bin: record ${i} (chunk ${c}, voxel ${v}) is out of range for the grid`);
    }
    // Composite key exact in float64: local_voxel_id < 2^16, chunk_id < 2^32.
    const key = c * 65536 + v;
    if (key <= previousKey) {
      throw new Error(`voxel_proxy.bin: record ${i} is out of (chunk_id, local_voxel_id) order or a duplicate`);
    }
    previousKey = key;

    chunkId[i] = c;
    localVoxelId[i] = v;
    count[i] = version === 1 ? view.getUint16(base + 6, true) : view.getUint32(base + 6, true);
    const colorOffset = version === 1 ? 8 : 10;
    for (let k = 0; k < 3; k++) colorRgb[i * 3 + k] = view.getUint8(base + colorOffset + k);
    flags[i] = view.getUint8(base + colorOffset + 3);

    if (runStart[c] < 0) runStart[c] = i;
    runEnd[c] = i + 1;
  }

  return { numVoxels, voxelsPerChunk, chunkId, localVoxelId, count, colorRgb, flags, runStart, runEnd };
}

/**
 * Fetches `voxel_proxy.bin` for a manifest and checks it describes exactly
 * this pack: the same grid, the record count the manifest advertises, and one
 * run per occupied chunk whose length is that chunk's `n_occupied_voxels` —
 * the property that makes "record i of the run == instance i of the chunk
 * mesh" true, which `VoxelProxyCloud` relies on to swap the two layers.
 */
export async function loadVoxelProxy(manifest: Manifest, signal?: AbortSignal): Promise<VoxelProxyData> {
  const entry = manifest.raw.voxel_proxy;
  const buffer = await fetchArrayBuffer(manifest.url(entry.path), signal);
  const data = parseVoxelProxy(buffer);

  if (data.numVoxels !== manifest.numVoxels || data.voxelsPerChunk !== manifest.voxelsPerChunk) {
    throw new Error(
      `voxel_proxy.bin grid ${data.numVoxels}/${data.voxelsPerChunk} disagrees with manifest ` +
        `${manifest.numVoxels}/${manifest.voxelsPerChunk}`,
    );
  }
  if (data.chunkId.length !== entry.n_voxels) {
    throw new Error(`voxel_proxy.bin has ${data.chunkId.length} records, manifest says ${entry.n_voxels}`);
  }
  let covered = 0;
  for (const chunk of manifest.chunks) {
    const start = data.runStart[chunk.chunk_id];
    const length = start < 0 ? 0 : data.runEnd[chunk.chunk_id] - start;
    if (length !== chunk.n_occupied_voxels) {
      throw new Error(
        `voxel_proxy.bin has ${length} records for chunk ${chunk.chunk_id}, ` +
          `manifest says ${chunk.n_occupied_voxels} occupied voxels`,
      );
    }
    covered += length;
  }
  if (covered !== data.chunkId.length) {
    throw new Error(
      `voxel_proxy.bin has ${data.chunkId.length - covered} records in chunks the manifest omits`,
    );
  }
  return data;
}
